firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -213,9 +217,11 @@ String getName() {
   @Override
   public Task<String> getId() {
     preConditionChecks();
-    Task<String> task = addGetIdListener();
-    backgroundExecutor.execute(this::doGetId);
-    return task;
+    synchronized (lockGetFid) {
+      TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+      taskCompletionSource.trySetResult(doGetId());
+      return taskCompletionSource.getTask();
+    }
```

Comment:
If we do the `cachedFid` solution, then it has to happen outside of the `synchronized` block of course.
The setting of `cachedFid` could happen anywhere where you think it makes sense.
```
if (null != cachedFid) {
  return cachedFid
}
synchronized (...)
```